hashes/src/sha256t.rs

```diff
@@ -6,11 +6,11 @@ use core::cmp;
 use core::marker::PhantomData;
 
 use crate::{sha256, FromSliceError, HashEngine as _};
-
-type HashEngine = sha256::HashEngine;
+#[cfg(doc)]
+use crate::sha256::Midstate;
 
 /// Trait representing a tag that can be used as a context for SHA256t hashes.
-pub trait Tag {
+pub trait Tag: Clone {
```

Comment:
This is breaking and shouldn't be needed. I'd rather impl `Clone` manually to get around `derive` restriction.

Reply:
I also thought about this, but I could go either way on it. If we remove the bound, then users who create their own tag types, without using the macro, need to remember to implement `Clone` on it or else they'll get downstream errors related to `Clone` not being implemented.

Maybe these errors will be clear enough, and this form of mistake rare enough, that it doesn't matter. I could go either way.

Reply:
> without using the macro, need to remember to implement `Clone` on it or else they'll get downstream errors related to `Clone` not being implemented on their hash types

No, I mean we implement `Clone` on `sha256t::Hash` for all tags. They'd only get such errors if they wanted to wrap the engine/hash somehow in a struct with generict tag *and use derive* to implement the traits. I think this is unlikely and even if it happens, manual trait impl should be fine.

> Though now that I think about this with fresh eyes, maybe this is already true...

It is for generic *tags*, for hash types it's not as long as they bound by `Hash` which already requires `Clone`.

Reply:
Ok, yeah, I think you're correct (and if not, this is at worst a papercut) and we can remove this.

Reply:
Can we do that as a follow up?

Tracked here: #4016